SwiftCompilerSources/Sources/Optimizer/Utilities/LifetimeDependenceUtils.swift

```diff
@@ -0,0 +1,505 @@
+      }
+    default:
+      break
+    }
+    // If the dependence chain has a phi, consider it a root. Dependence roots
+    // are currently expected to dominate all dependent values.
+    if Phi(value) != nil {
+      return introducer(value)
+    }
+    return walkUpDefault(forwarded: value)
+  }
+}
+
+/// Walk down dependent values.
+///
+/// Delegates value def-use walking to the ForwardingUseDefWalker and
+/// overrides copy, move, borrow, and mark_dependence.
+///
+/// Stops at an address value. Finding the leaf uses of an address
+/// requires an AddressDefUseWalker.
+///
+/// Ignores trivial values (~Escapable types are never
+/// trivial. Escapable types may only be lifetime-depenent values if
+/// they are non-trivial).
+///
+/// Skips uses within nested borrow scopes.
+///
+/// TODO: override BeginAccessInst to handle EndAccessInst.
+///
+/// Minimal requirements:
+///   needWalk(for value: Value) -> Bool
+///   leafUse(_ operand: Operand) -> WalkResult
+///   deadValue(_ value: Value, using operand: Operand?) -> WalkResult
+///
+/// Start walking:
+///   walkDown(root: Value)
+protocol LifetimeDependenceDefUseWalker : ForwardingDefUseWalker {
+  var function: Function { get }
+  var context: Context { get }
+
+  // Minimally, check a ValueSet. This walker may traverse chains of
+  // aggregation and destructuring by default. Implementations may
+  // handle phis.
+  mutating func needWalk(for value: Value) -> Bool
+
+  mutating func leafUse(_ operand: Operand) -> WalkResult
+
+  // Report any initial or dependent value with no uses. Only relevant for
+  // guaranteed values or incomplete OSSA. This could be a dead
+  // instruction, a terminator in which the result is dead on one
+  // path, or a dead phi.
+  //
+  // \p operand is nil if \p value is the root.
+  mutating func deadValue(_ value: Value, using operand: Operand?) -> WalkResult
+
+  mutating func walkDownUses(of: Value, using: Operand?) -> WalkResult
+    
+  mutating func walkDown(operand: Operand) -> WalkResult
+}
+
+extension LifetimeDependenceDefUseWalker {
+  // Callback from ForwardingDefUseWalker.
+  mutating func walkDown(operand: Operand) -> WalkResult {
+    switch operand.instruction {
+    case let copy as CopyValueInst:
+      return walkDownUses(of: copy, using: operand)
+
+    case let move as MoveValueInst:
+      return walkDownUses(of: move, using: operand)
+
+    case let borrow as BeginBorrowInst:
+      // Skip the uses within nested borrow scopes.
+      for endOperand in BeginBorrowValue(borrow)!.scopeEndingOperands {
+        switch endOperand.instruction {
+        case is EndBorrowInst:
+          if leafUse(endOperand) == .abortWalk {
+            return .abortWalk
+          }
+        case is BranchInst:
+          let phi = Phi(using: endOperand)!
+          if walkDownUses(of: phi.value, using: endOperand) == .abortWalk {
+            return .abortWalk
+          }
+        default:
+          fatalError("unknown borrow scope end")
```

Comment:
Worth encoding these two possibilities as an enum which `scopeEndingOperands` is a sequence of?

Reply:
I encoded EndBorrowOperand as an enum. But there are always unreachable cases here because borrow scopes can now counterintuitively be owned values in the case of partial_apply.